lib/server/parse-websocket.js

```diff
@@ -6,33 +6,51 @@ function parseWebSocketRequest (socket, opts, params) {
   if (!opts) opts = {}
   params = JSON.parse(params) // may throw
 
-  params.action = common.ACTIONS.ANNOUNCE
   params.type = 'ws'
   params.socket = socket
+  if (params.action === 'announce' || params.answer || params.offers) {
+    params.action = common.ACTIONS.ANNOUNCE
 
-  if (typeof params.info_hash !== 'string' || params.info_hash.length !== 20) {
-    throw new Error('invalid info_hash')
-  }
-  params.info_hash = common.binaryToHex(params.info_hash)
+    if (typeof params.info_hash !== 'string' || params.info_hash.length !== 20) {
+      throw new Error('invalid info_hash')
+    }
+    params.info_hash = common.binaryToHex(params.info_hash)
 
-  if (typeof params.peer_id !== 'string' || params.peer_id.length !== 20) {
-    throw new Error('invalid peer_id')
-  }
-  params.peer_id = common.binaryToHex(params.peer_id)
+    if (typeof params.peer_id !== 'string' || params.peer_id.length !== 20) {
+      throw new Error('invalid peer_id')
+    }
+    params.peer_id = common.binaryToHex(params.peer_id)
 
-  if (params.answer) {
-    if (typeof params.to_peer_id !== 'string' || params.to_peer_id.length !== 20) {
-      throw new Error('invalid `to_peer_id` (required with `answer`)')
+    if (params.answer) {
+      if (typeof params.to_peer_id !== 'string' || params.to_peer_id.length !== 20) {
+        throw new Error('invalid `to_peer_id` (required with `answer`)')
+      }
+      params.to_peer_id = common.binaryToHex(params.to_peer_id)
     }
-    params.to_peer_id = common.binaryToHex(params.to_peer_id)
-  }
 
-  params.left = Number(params.left) || Infinity
-  params.numwant = Math.min(
-    Number(params.offers && params.offers.length) || 0, // no default - explicit only
-    common.MAX_ANNOUNCE_PEERS
-  )
-  params.compact = -1 // return full peer objects (used for websocket responses)
+    params.left = Number(params.left) || Infinity
+    params.numwant = Math.min(
+      Number(params.offers && params.offers.length) || 0, // no default - explicit only
+      common.MAX_ANNOUNCE_PEERS
+    )
+    params.compact = -1 // return full peer objects (used for websocket responses)
+  } else if (params.action === 'scrape') {
+    params.action = common.ACTIONS.SCRAPE
+
+    if (typeof params.info_hash === 'string') params.info_hash = [ params.info_hash ]
+    if (Array.isArray(params.info_hash)) {
+      params.info_hash = params.info_hash.map(function (binaryInfoHash) {
+        if (typeof binaryInfoHash !== 'string' || binaryInfoHash.length !== 20) {
+          throw new Error('invalid info_hash')
+        }
+        return common.binaryToHex(binaryInfoHash)
+      })
+    } else {
+      params.info_hash = common.binaryToHex(params.info_hash)
```

Comment:
This else statement can be removed. It should never be true. If `params.info_hash` is a string it will be converted to an array with a single string in it on line 40.